### PR TITLE
Construct de Rham sequence with a NURBS/spline mapping

### PR DESCRIPTION
**Pull request by vcarlier**
Solve #243:

* In the class method `Geometry.from_discrete_mapping` set the `SplineMapping` as the `callable_mapping` of the symbolic mapping before calling the constructor of `Geometry`.
* Modify the constructor of class `Pushforward` to accept a symbolic mapping having a `SplineMapping` as its callable mapping. Accordingly, remove the assert which was requiring the mapping to be analytic.
* Add some tests is `psydac.api.tests.test_build_derham_mapping.py` to prove that this method works.

### CODE REVIEW

File: psydac/api/postprocessing.py
Comment (yguclu) [inline psydac/api/postprocessing.py:1]: Can you please check if the `mapping` argument can be of type `SplineMapping`? If not, we could simplify the code a bit.

File: psydac/feec/pushforward.py
Comment (yguclu) [inline psydac/feec/pushforward.py:1]: Also here. Can you please check if the `mapping` argument can be of type `SplineMapping`? If not, we could simplify the code a bit.

File: psydac/api/postprocessing.py
Comment (vcarlier) [reply psydac/api/postprocessing.py:1]: Apparently it's usefull in the multipatch setting where the mapping of each patch could be represented as a SplineMapping if I understood correctly.

File: psydac/feec/pushforward.py
Comment (vcarlier) [reply psydac/feec/pushforward.py:1]: Same : Apparently it's usefull in the multipatch setting where the mapping of each patch could be represented as a `SplineMapping` if I understood correctly.

**Review by yguclu — CHANGES_REQUESTED** [review]

File: psydac/feec/pushforward.py [hunk 1]
```diff
@@ -107,11 +107,16 @@ def __init__(
         if isinstance(mapping, Mapping):
             self._mesh_grids = np.meshgrid(*grid_local, indexing='ij', sparse=True)
             # No support for non analytical mappings for now
```
[the remark this answers is not in the document]
Comment (vcarlier) [reply psydac/feec/pushforward.py]: Done in [fb754be](https://github.com/pyccel/psydac/pull/366/commits/fb754be254176f624e2730bec5e89509a5a2134c)

File: psydac/cad/geometry.py [hunk 2]
```diff
@@ -122,20 +122,43 @@ def __init__(self, domain=None, ncells=None, periodic=None, mappings=None,
     # Option [2]: from a discrete mapping
     #--------------------------------------------------------------------------
     @classmethod
-    def from_discrete_mapping(cls, mapping, comm=None):
-        """Create a geometry from one discrete mapping."""
+    def from_discrete_mapping(cls, mapping, comm=None, name= ''):
+        """Create a geometry from one discrete mapping.
+        Parameters
```
[the remark this answers is not in the document]
Comment (vcarlier) [reply psydac/cad/geometry.py]: Done in [fb754be](https://github.com/pyccel/psydac/pull/366/commits/fb754be254176f624e2730bec5e89509a5a2134c)

File: psydac/cad/geometry.py [hunk 3]
```diff
@@ -122,20 +122,43 @@ def __init__(self, domain=None, ncells=None, periodic=None, mappings=None,
...
+        ----------
+        mapping : SplineMapping
+            The Mapping from the unit square to to physical domain.
```
[the remark this answers is not in the document]
Comment (vcarlier) [reply psydac/cad/geometry.py]: Done in [fb754be](https://github.com/pyccel/psydac/pull/366/commits/fb754be254176f624e2730bec5e89509a5a2134c)

File: psydac/api/tests/test_build_derham_mapping.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    os.remove("./export_sol.h5")
+    os.remove("./fields_test.0000.vtu")
+
+if __name__ == '__main__':
+    #test_build_derham_spline_mapping_id_1d([2], [10], [True]) 
+    import time   
+    test_build_derham_spline_mapping_id_2d([2,2], [5,10], [True,True])  
+    #time.sleep(15) 
+    test_build_derham_spline_mapping_id_2d([2,2], [10,10], [True,True])    
+    #test_build_derham_spline_mapping_id_3d([2,3,4], [10,11,3], [True,False,False])
```
[the remark this answers is not in the document]
Comment (vcarlier) [reply psydac/api/tests/test_build_derham_mapping.py]: Done in [fb754be](https://github.com/pyccel/psydac/pull/366/commits/fb754be254176f624e2730bec5e89509a5a2134c)

File: psydac/cad/geometry.py [hunk 5]
```diff
@@ -122,20 +122,43 @@ def __init__(self, domain=None, ncells=None, periodic=None, mappings=None,
...
+        comm : MPI.Comm
+            MPI intra-communicator.
+            
+        name : string
+            Optional name for the Mapping that will be created. 
+            Needed to avoid conflicts in case several mappings are created
+        """
+
+        if mapping.ldim == 1:
+            M        = Mapping('mapping' + name, dim=1)
```
[the remark this answers is not in the document]
Comment (vcarlier) [reply psydac/cad/geometry.py]: Done in [fb754be](https://github.com/pyccel/psydac/pull/366/commits/fb754be254176f624e2730bec5e89509a5a2134c). 

It was probably just a default name but I have no clue why `_0`.

**Review by yguclu — APPROVED** [review]

**Review by yguclu — APPROVED** [review]

**Pull request merged**